### PR TITLE
2.0-preview branch

### PR DESCRIPTION
**Pull request by bicknellr**
~~Needs Polymer/polymer#3923.~~

### CODE REVIEW

File: test/activate-event.html [hunk 1]
```diff
@@ -49,8 +49,13 @@
 
       var s;
 
-      setup(function () {
+      setup(function(done) {
         s = fixture('test');
+
+        s.addEventListener('iron-items-changed', function onIronItemsChanged() {
```
Comment (e111077) [inline test/activate-event.html]: Why is this necessary in setup? Is there a breaking change in Polymer or in iron-selector? Why will `forceSynchronousItemUpdate` not work here?

File: iron-selectable.html [hunk 2]
```diff
@@ -275,8 +305,8 @@
     },
 
     _updateAttrForSelected: function() {
-      if (this._shouldUpdateSelection) {
-        this.selected = this._indexToValue(this.indexOf(this.selectedItem));
+      if (this.selectedItem) {
```
Comment (bicknellr) [inline iron-selectable.html]: `selectedItem` should be a reference to the node itself, rather than its index (`selected`, as you mentioned above).

File: iron-multi-selectable.html [hunk 3]
```diff
@@ -12,7 +12,15 @@
 <link rel="import" href="iron-selectable.html">
 
 <script>
-  /** @polymerBehavior Polymer.IronMultiSelectableBehavior */
+  /**
+   * ### Notable breaking changes between 1.x and 2.x (hybrid):
```
Comment (bicknellr) [inline iron-multi-selectable.html]: This is here so that it shows up in the docs shown by iron-component-page when you switch to `IronSelectableBehavior` or `IronMultiSelectableBehavior`, but I think adding a reference to it in the README would be good.

File: iron-selectable.html [hunk 4]
```diff
@@ -13,7 +13,29 @@
 
 <script>
 
-  /** @polymerBehavior */
+  /**
+   * ### Notable breaking changes between 1.x and 2.x (hybrid):
```
Comment (bicknellr) [inline iron-selectable.html]: ([same as above](https://github.com/PolymerElements/iron-selector/pull/139/files#r110283232))

File: iron-multi-selectable.html (hunk 3 above)
Comment (bicknellr) [inline iron-multi-selectable.html]: Huh, it seems like the README is meant to be updated with stuff from these files anyways.

Comment (e111077) [inline iron-multi-selectable.html]: I can go either way on this. lmk what you end up choosing

File: test/activate-event.html (hunk 1 above)
Comment (bicknellr) [inline test/activate-event.html]: This is because of a breaking change that's a combined effort of IronSelectableBehavior, Polymer, and the custom element reaction queues introduced in v1. I left a really vague note about this [here](https://github.com/PolymerElements/iron-selector/pull/139/files#diff-c64eedc3daead72759c9014ab0eb08c4R19). The gist is that IronSelectableBehavior can't update its items and selected states synchronously within `attached` without breaking its most basic use case.

Specifically, if you have an element with IronSelectableBehavior with an `attr-for-selected` value and an item (1) with a property that reflects to that same attribute referenced by `attr-for-selected`, (2) this property has a default value of `false`, and (3) this item will be the selected item; then you end up with this lame series of events:
  1. The tree is cloned / constructed and the default values are stored by Polymer in some internal object.
  1. The tree is connected and `connectedCallback` reactions are enqueued for the selector and the item.
  1. The selector's `connectedCallback` is dequeued and run first, during which it calls `setAttribute` on the item.
  1. The item's `connectedCallback` is dequeued and run *as the next frame on the stack below the function calling `setAttribute`* which then reads the currently known attribute value out of the internal object and reflects it by calling `removeAttribute` because the default is `false` and `attributeChangedCallback` hasn't yet been called to indicate otherwise.
  1. The item's *first* pending `attributeChangedCallback` reaction is called (enqueued by `setAttribute`), setting the property to `true`.
  1. The item's *second* pending `attributeChangedCallback` reaction is called (enqueued by `removeAttribute`), setting the property back to `false`.

So, the selector ends up trying to call `setAttribute` on an item which already has a pending `connectedCallback` and that `connectedCallback` ends up invalidating the `setAttribute` call by reflecting that property's default value back to the attribute (i.e. removing it). This exact scenario is what happens with `paper-radio-group`, `paper-radio-button` and the `checked` attribute.

`forceSynchronousItemUpdate` works here as well but I'd rather encourage people that run across this to use `iron-items-changed`.

Comment (bicknellr) [inline test/activate-event.html]: Oh, so that didn't actually explain the fix... So the most reliable fix seemed to be to *only* update items and selection when the `observeNodes` callback is called, which means that the tests need to wait for `iron-items-changed` to be sure that `items` and other things that depend on `items` have been updated.

Comment (bicknellr) [inline test/activate-event.html]: Also, here's a more minimal repro: http://jsbin.com/quvetix/edit?html,output

File: test/basic.html [hunk 5]
```diff
@@ -217,42 +234,47 @@
...
+          document.body.appendChild(selector);
 
+          // Create and append a new item which should become selected.
+          var child = document.createElement('div');
+          child.textContent = 'Item 0';
           Polymer.dom(selector).appendChild(child);
-          document.body.appendChild(selector);
 
-          Polymer.Base.async(function() {
+          selector.addEventListener('iron-items-changed', function onIronItemsChanged() {
```
Comment (bicknellr) [inline test/basic.html]: No, because the `iron-items-changed` event is only fired in the callback for the MutationObserver (`observeNodes`), which doesn't happen synchronously during the `appendChild` above.

File: test/activate-event.html (hunk 1 above)
Comment (e111077) [inline test/activate-event.html]: Seems good. The main reason I ask is because we asked a lot of users to use the `forceSynchronousItemUpdate` as a fix to making iron-selector asynchronous which caused an overwhelming amount of blowback from our users.

File: test/basic.html (hunk 5 above)
Comment (e111077) [inline test/basic.html]: 👍

File: iron-selectable.html (hunk 2 above)
Comment (e111077) [inline iron-selectable.html]: haha didn't see that 😵

**Comment by bicknellr** [comment]
I need to update the README here since it's not being updated automatically anymore.
